lib/ecto/association.ex

```diff
@@ -507,7 +507,7 @@ defmodule Ecto.Association.Has do
   def joins_query(%{queryable: queryable, related_key: related_key,
                     owner: owner, owner_key: owner_key}) do
     from o in owner,
-      join: q in ^queryable,
```

Comment:
I am afraid this is not the correct place to fix this. Somewhere internally we handle `join: q in Foo` and we need to make sure the conversion happens there.